### PR TITLE
Synchronising ParseNumber & MathChars with the new CoreCLR versions

### PR DESCRIPTION
**Pull request by varocarbas**
Updating the `ParseNumber`/`MatchChars` contents with the new optimised versions [already merged to CoreCLR](https://github.com/dotnet/coreclr/pull/3163/).

### CODE REVIEW

**Comment by stephentoub** [comment]
Do you know where in corefx this code is used?

**Comment by varocarbas** [comment]
@stephentoub At least, by `BigInteger`. Additionally, this file is identical to one in CoreRT which is already being updated and I guess that have the same version of the same methods everywhere is a good thing.

**Comment by stephentoub** [comment]
Thanks.

cc: @mellinoe

**Comment by varocarbas** [comment]
@stephentoub You are welcome.

**Comment by tarekgh** [comment]
CC @jkotas

**Comment by tarekgh** [comment]
I know @jkotas reviewed similar change in corert but I am seeing this change could be clean up more than optimization. right?

**Comment by tarekgh** [comment]
LGTM. 

@mellinoe please have a look too?

**Comment by jkotas** [comment]
> I am seeing this change could be clean up more than optimization. right?

Yes. Ideally, the CoreRT and CoreFX versions of the file would be 100% identical.

File: src/Common/src/System/Globalization/FormatProvider.Number.cs [hunk 1]
```diff
@@ -382,86 +372,80 @@ private unsafe static bool ParseNumber(ref char* str, NumberStyles options, ref
...
+                            // currSymbol to NULL so that we won't search it again in the later code path.
+                            p = next - 1;
+                        }
+                        else 
+                        {
+                            break;
+                        }
                     }
-                    ch = *++p;
+                    ch = *++p;
```
Comment (mellinoe) [inline src/Common/src/System/Globalization/FormatProvider.Number.cs]: nit: a lot of lines have extra spaces at the end, can you clean those up?

**Comment by mellinoe** [comment]
>  ideally, the CoreRT and CoreFX versions of this file would be 100% identical.

Yes, that would be nice. I looked through the diffs between this and corert's, and mostly the differences are stylistic, but in a couple of cases there are functional differences (it looks like corert is doing extra comparisons that are pointless to my eye).

LGTM, although I'd like to see the couple of stray spaces cleaned up.

**Comment by varocarbas** [comment]
@mellinoe @tarekgh @jkotas Note that the CoreRT version is just a mere copy (a plain copy-paste action executed by a person different than me, the author of these modifications). The original version is [the CoreCLR one](https://github.com/dotnet/coreclr/pull/3163/). You can read the comments there to undersand what this is about. Also note that I documented in detailed everything in [varocarbas.com Project 7](http://varocarbas.com/open_net/) (some updates pending though).

**Comment by varocarbas** [comment]
@mellinoe  I will remove the blank spaces right away.

Regarding your
> mostly the differences are stylistic

I am afraid that this isn't true. The origin of the modifications was adapting the code to a version closer to what the compiler generates (analysis via decompiler), these modifications proved to deliver a slightly better performance (take a look at the original CoreCLR conversation to understand that I did all the tests in a not-that-powerful-computer where the differences I observed were notably smaller than the ones in more powerful computers). I also removed various `els`e statements (this bit was more influential than the previous one) and some redundant variables (the most relevant factor). You can read more detailed explanations about all this [here](http://varocarbas.com/open_net/improvements_final/). 

These improvements are more influential for decimal parsing, mainly for the `decimal` type (which has its own type of parsing), but all the types are improved. I can do some specific tests for `BigInteger`/integer types to show the real gain here. I will do it by next week.

File: src/Common/src/System/Globalization/FormatProvider.Number.cs (hunk 1 above)
Comment (varocarbas) [inline src/Common/src/System/Globalization/FormatProvider.Number.cs]: OK.

**Comment by varocarbas** [comment]
@mellinoe Sorry for not having got your point right. In any case, this was a good excuse to highlight some issues which might not be evident to everyone.

**Pull request merged**